### ISSUE TITLE
Throw on 4XX / 5XX

### ISSUE THREAD

**Issue by jackfranklin** [jackfranklin/fetch-factory#5]
From: https://github.com/github/fetch#caveats

> The Promise returned from fetch() won't reject on HTTP error status even if the response is a HTTP 404 or 500. Instead, it will resolve normally, and it will only reject on network failure, or if anything prevented the request from completing.

We should define a second default `response` interceptor that checks the status and throws, as I think it's a useful behaviour that should be provided by default. Users can always override.

**Comment by majodev** [comment]
+1, furthermore it would be really useful to allow the definition of request & response interceptors on a per method basis.

**Comment by jackfranklin** [comment]
@majodev can you raise a separate issue for that please so it doesn't get lost?

**Comment by jackfranklin** [comment]
So for this I think we basically want to take the example from the whatwg/fetch readme:

``` js
function throwOnResponseError(response) {
  if (response.status >= 200 && response.status < 300) {
    return response
  } else {
    var error = new Error(response.statusText)
    error.response = response
    throw error
  }
}
```

And define that as the first default response interceptor, which would be added here: https://github.com/jackfranklin/fetch-factory/blob/master/src/index.js#L47

We'd then need to update the README to document that this is the default behaviour, and also that if people override the response interceptors they will lose this functionality.

We might want to also export the `throwOnResponseError` so that people can use it, for example:

``` js
import fetchFactory, { throwOnResponseError } from 'fetch-factory';

const UserFactory = fetchFactory.create({
  url: 'http://www.api.com/users',
  method: 'GET',
  interceptors: {
    response: [ throwOnResponseError, (data) => ({ name: 'bob' }) ],
  },
}, {
  findAll: {},
});
```

**Comment by ArnaudRinquin** [comment]
I think this functionality does not belong in this library as error handling can be very business rules specific (handle `50X` differently than `40X`, etc...).

I wrote a very small composable library [`promise-reject-status-above`](https://github.com/songkick/promise-reject-status-above) that handle that case without creating coupling or adding weight to a specific fetch helper / library.

You can basically just wrap your fetch promise on it.

From the lib example:

``` js
var rejectStatusAbove = require('promise-reject-status-above');
var rejectAbove400 = rejectStatusAbove({status: 400});

function fetch400(){
    return fetch('/path/to/a/http-400-bad-request');
}

rejectAbove400(fetch400)()
  .then(function(response){
    // should not happen here
  }).catch(function(err){
      // err instanceof rejectStatusAbove.StatusAboveError === true
      // err === {
      //   message: 'Response status above accepted status',
      //   settings: {
      //     status: 400,
      //   },
      //   fn: fetch400,
      //   response: window.Response // the original fetch Response
      // }
  });
```

Compose :all-the-things:

**Comment by jackfranklin** [comment]
@ArnaudRinquin I think I disagree with you here - right now I find this behaviour of the fetch API not intuitive:

``` js
fetch('some400url').then(() => {
  // this got called
}).catch(() => {
  // but I'd expect this to
});
```

People can deal with the errors in their own way in the `catch`, or over-ride the interceptors. I think this makes sense as default behaviour as long as it's documented how to change it.

**Comment by jackfranklin** [comment]
PS: I believe @iam-peekay is working on this issue, so no one else take it please else we'll have wasted effort!

**Comment by ArnaudRinquin** [comment]
@jackfranklin I agree that fetch API is a little bit frustrating but it makes sense as HTTP status and request success are not exactly the same.

The reason why I wrote `promise-reject-status-above` is that I needed the simplest possible helper to reject fetch promises based on HTTP status without coupling myself to an _API framework_.

I initially though about writing something similar to `fetch-factory` (which I did in a different flavour with [`fetch-decorators`](https://github.com/ArnaudRinquin/fetch-decorators)), but I found that writing many promise wrappers ([`retry`](https://github.com/songkick/promise-retry), [`timeout`](https://github.com/songkick/promise-timeout), [`reject-status-above`](https://github.com/songkick/promise-reject-status-above)) would be way more modular + easier to test + lower the overall coupling between things.

TL;DR: small libraries over one-do-all _frameworky_ library.

But that's really just my two cents.

**Comment by jackfranklin** [comment]
@ArnaudRinquin we might have to agree to disagree ;) I completely understand your POV and can see both sides, I think for me this little tweak makes sense - I'm still for adding it and very clearly documenting that it is a thing and how to stop it being a thing.

**Comment by ArnaudRinquin** [comment]
> we might have to agree to disagree ;)

1000 \* :+1: !

**Comment by iam-peekay** [comment]
@jackfranklin just so it's clear to me that i understand this correctly, is there a default response interceptor defined already? is "removeNullFetchOptions" a default response interceptor?

**Comment by jackfranklin** [comment]
@iam-peekay ah sorry, the code has changed since that link!

Here: https://github.com/jackfranklin/fetch-factory/blob/master/src/index.js#L53, is the line with the array of default response interceptors - we want to add a new one to the beginning.

**Comment by jackfranklin** [comment]
Closed by #19
